lib/Conversion/TorchOnnxToTorch/OnnxRnnExpander.cpp

```diff
@@ -0,0 +1,312 @@
+    Block *loopBody =
+        b.createBlock(&loop.getRegion(), loop.getRegion().begin(),
+                      TypeRange({
+                          loopIndexType,
+                          yTy,
+                          hTy,
+                      }),
+                      {loc, loc, loc} // locs for the loop body arguments
+        );
+
+    Value loopIndex = loopBody->getArgument(0);
+    Value Y_prev = loopBody->getArgument(1);
+    Value H_prev = loopBody->getArgument(2);
+
+    auto xTy = cast<ValueTensorType>(X.getType());
+    auto XtType = b.getType<ValueTensorType>(
+        llvm::SmallVector<int64_t>{batch_size, input_size}, xTy.getDtype());
+
+    Value Xt = b.create<AtenSelectIntOp>(XtType, X, cstZero, loopIndex);
+
+    Value H_new = rnn_cell(b, Xt, H_prev, weights, activations);
+
+    Type hTyUnsqueezed = b.getType<ValueTensorType>(
+        llvm::SmallVector<int64_t>{1, batch_size, hidden_size}, hTy.getDtype());
+    Value H_new_unsqueezed =
+        b.create<AtenUnsqueezeOp>(hTyUnsqueezed, H_new, cstZero);
+
+    auto loopIndexPlusOne = b.create<AtenAddIntOp>(intType, loopIndex, cstOne);
+    Value Y_new =
+        b.create<AtenSliceScatterOp>(yTy, Y_prev, H_new_unsqueezed, cstZero,
+                                     loopIndex, loopIndexPlusOne, cstOne);
+
+    b.create<PrimLoopConditionOp>(loopConditionTrue,
+                                  ValueRange({Y_new, H_new}));
+  }
+  RnnLayerOutput output;
+  output.Y = loop.getResult(0);
+  output.Y_h = loop.getResult(1);
+  return output;
+}
+
+LogicalResult OnnxRnnExpander(OpBinder binder,
+                              ConversionPatternRewriter &rewriter) {
+  Location loc = binder.getLoc();
+  mlir::ImplicitLocOpBuilder b(loc, rewriter);
+
+  std::string direction;
+
+  ValueTensorType yTy, Y_hType;
+  if (binder.tensorResultTypeAtIndex(yTy, 0) ||
+      binder.tensorResultTypeAtIndex(Y_hType, 1)) {
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "At least one output must be present");
+  }
+  Value X;
+  if (binder.tensorOperandAtIndex(X, 0))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor X");
+  Value W;
+  if (binder.tensorOperandAtIndex(W, 1))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor W");
+  Value R;
+  if (binder.tensorOperandAtIndex(R, 2))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor R");
+  int64_t hidden_size;
+  if (binder.s64IntegerAttr(hidden_size, "hidden_size"))
+    return rewriter.notifyMatchFailure(
+        binder.op, "Missing required attribute hidden_size");
+
+  auto xTy = cast<ValueTensorType>(X.getType());
+  auto wTy = cast<ValueTensorType>(W.getType());
+  Value B;
+  if (binder.tensorOperandAtIndex(B, 3)) {
+    B = b.create<AtenZerosOp>(W.getType(), W);
```

Comment:
From the docs, B has shape `[num_directions, 2*hidden_size]` and W has shape `[num_directions, hidden_size, input_size]`, so I'm not sure I understand why `W` is being used to create `B` here.

Reply:
Good catch. Fixing now.